### PR TITLE
fix: SFDX: Rename Component now shows an error message if one uses an existing file name as the new component name to avoid data loss.

### PR DESCRIPTION
**Pull request by floralan**
### What does this PR do?
Fix the bug that SFDX: Rename Component can silently replace a file if one uses an existing file name as new component name. 

### What issues does this PR fix or reference?
#4024, @W-10983487@

### Functionality Before
if one renames a lightning/aura component and there is an existing file with the same name, it will be replaced without warning. 

### Functionality After
Added an error message saying that: "the component name is already in use for a file under current component directory" to prevent user from using an existing file name as new component name to avoid file loss.

### CODE REVIEW

**Review by sbudhirajadoc — CHANGES_REQUESTED** [review]
Change -- Component name is already in use for a file in current component directory.

To. --  This component name is already in use in the current component directory. Choose a different name and try again.

File: packages/salesforcedx-vscode-core/src/messages/i18n.ts [hunk 1]
```diff
@@ -624,6 +624,7 @@ export const messages = {
   sobject_refresh_standard: 'Standard SObjects',
   force_rename_lightning_component: 'SFDX: Rename Component',
   rename_component_input_dup_error: 'Component name is already in use in LWC or Aura',
+  rename_component_input_dup_file_name_error: 'This component name is already in use in the current component directory. Choose a different name and try again.',
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/messages/i18n.ts:627]: nit: should this be: This file name ?

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceRenameLightningComponent.test.ts [hunk 2]
```diff
@@ -213,6 +214,48 @@ describe('Force Rename Lightning Component', () => {
...
+      try {
+        const executor = new RenameLwcComponentExecutor(sourceUri.fsPath);
+        await executor.run({
+          type: 'CONTINUE',
+          data: {name: 'templateOne'}
+        });
+      } catch (e) {
+        exceptionThrown = true;
+      }
+      expect(exceptionThrown).to.equal(true);
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceRenameLightningComponent.test.ts:235]: nit: you could verify what exception is getting thrown here. I've def had tests in the past where I thought it was throwing one kind of error, but it was actually failing for a different reason. same on line 256.

File: packages/salesforcedx-vscode-core/src/messages/i18n.ts (hunk 1 above)
Comment (gbockus-sf) [reply packages/salesforcedx-vscode-core/src/messages/i18n.ts:627]: nit^2: should we include the file name?

File: packages/salesforcedx-vscode-core/src/commands/forceRenameLightningComponent.ts [hunk 3]
```diff
@@ -154,6 +156,29 @@ async function isDuplicate(componentPath: string, newName: string): Promise<bool
   return (allLwcComponents.includes(newName) || allAuraComponents.includes(newName));
 }
 
+// enforce unique new file name under current component directory to avoid file loss
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/forceRenameLightningComponent.ts:159]: nit: for functions should we use the ```
/** 
 *
 **/```
comment syntax?

**Review by gbockus-sf — COMMENTED** [review]
good stuff.  some nits to consider

File: packages/salesforcedx-vscode-core/src/messages/i18n.ts (hunk 1 above)
Comment (sbudhirajadoc) [reply packages/salesforcedx-vscode-core/src/messages/i18n.ts:627]: OK to switch to "This file name is already in use in the current component directory...."  if this is actually a file name in a component directory.

**Review by sbudhirajadoc — APPROVED** [review]

**Comment by floralan** [comment]
> good stuff. some nits to consider

Great suggestions! Have updated the code based on your comments.

File: packages/salesforcedx-vscode-core/src/commands/forceRenameLightningComponent.ts [hunk 4]
```diff
@@ -154,6 +156,31 @@ async function isDuplicate(componentPath: string, newName: string): Promise<bool
...
+ * check duplicate name under current component directory and __tests__ directory to avoid file loss
+ */
+async function checkForDuplicateInComponent(componentPath: string, newName: string, items: string[]) {
+  let allFiles = items;
+  if (items.includes(TEST_FOLDER)) {
+    const testFiles = await fs.promises.readdir(path.join(componentPath, TEST_FOLDER));
+    allFiles = items.concat(testFiles);
+  }
+  // get all file names and filter out folder names
+  const allFileNames = allFiles.map(file => {
```
Comment (randi274) [inline packages/salesforcedx-vscode-core/src/commands/forceRenameLightningComponent.ts:169]: Is it possible to refactor this to create a named function so the comment can be removed?

Comment (floralan) [reply packages/salesforcedx-vscode-core/src/commands/forceRenameLightningComponent.ts:169]: @randi274 Sounds great! updated the code.

**Review by randi274 — APPROVED** [review]
**QA Complete**
* Can rename LWC component without an duplicate names → ✅
* Can rename LWC component w/ another component having a JS file with the intended name →   ✅
* Can’t rename LWC component with another LWC component bundle having the intended name → ✅
* Can’t rename LWC component w/ Aura having the intended bundle name → ✅
* Can’t rename Aura component w/ LWC having the intended bundle name → ✅
* Can’t rename LWC component that has a .js file (or any other file) matching the new intended name ✅
* Can’t rename LWC component that has a .js file (or any other file) matching the new intended name ✅
* Can’t run the command from the command palette at all ✅

I did a little extra QA since I didn't get a chance to on initial pass through. I also encountered a few additional questions about the feature, neither of which should prevent merge of this fix.

* ❓I noticed `SFDX: Rename Component` was not available on the __tests__ directory or any files therein. Should it be? 
* ❓Is it possible to reopen the new component after renaming? When I run the default Rename it does reopen the bundle, and currently functionality is that focus is lost and I have to find the component again.

**Comment by floralan** [comment]
> **QA Complete**
> 
> * Can rename LWC component without an duplicate names → ✅
> * Can rename LWC component w/ another component having a JS file with the intended name →   ✅
> * Can’t rename LWC component with another LWC component bundle having the intended name → ✅
> * Can’t rename LWC component w/ Aura having the intended bundle name → ✅
> * Can’t rename Aura component w/ LWC having the intended bundle name → ✅
> * Can’t rename LWC component that has a .js file (or any other file) matching the new intended name ✅
> * Can’t rename LWC component that has a .js file (or any other file) matching the new intended name ✅
> * Can’t run the command from the command palette at all ✅
> 
> I did a little extra QA since I didn't get a chance to on initial pass through. I also encountered a few additional questions about the feature, neither of which should prevent merge of this fix.
> 
> * ❓I noticed `SFDX: Rename Component` was not available on the **tests** directory or any files therein. Should it be?
> * ❓Is it possible to reopen the new component after renaming? When I run the default Rename it does reopen the bundle, and currently functionality is that focus is lost and I have to find the component again.

@randi274 Thank you for the very detailed QA report! For your first questions, I changed the code to enable rename command for **__tests__** folder and any .js files under it. Could you take time to verify you can see the change on your end as well? For the second question, I think that's a very good feature users may expect. We can create a user story for it!

**Comment by randi274** [comment]
Thanks for making that change so quickly @floralan! I think it might be a bit more of a complicated story, as it doesn't look like it's working as I'd expect quite yet; can you revert that commit, and then we'll add another user story to address that one?

**Comment by floralan** [comment]
> 
Yes, making another story for the changes sounds a better plan! I just reset the branch, once all the tests done, we can get this PR merged 👍

**Pull request merged**